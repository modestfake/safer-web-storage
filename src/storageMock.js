export default function getStorage({ errorMessage }) {
  return {
    getItem() {
      console.warn(errorMessage)
    },
    setItem() {
      console.warn(errorMessage)
    },
    removeItem() {
      console.warn(errorMessage)
    },
    clear() {
      console.warn(errorMessage)
    },
    get length() {
      console.warn(errorMessage)

      return 0
    },
  }
}
