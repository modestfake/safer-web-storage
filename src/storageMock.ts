/* eslint-disable @typescript-eslint/no-unused-vars */
import { Options } from './SafeStorage'

class MockedStorage implements Storage {
  errorMessage: string

  constructor(options: Required<Options>) {
    this.errorMessage = options.errorMessage
  }

  getItem(key: string) {
    console.warn(this.errorMessage)

    return null
  }

  setItem(key: string) {
    console.warn(this.errorMessage)

    return null
  }

  removeItem(key: string) {
    console.warn(this.errorMessage)

    return null
  }

  key() {
    console.warn(this.errorMessage)

    return null
  }

  clear() {
    console.warn(this.errorMessage)

    return null
  }

  get length() {
    console.warn(this.errorMessage)

    return 0
  }
}

export default function (options: Required<Options>) {
  return new MockedStorage(options)
}
