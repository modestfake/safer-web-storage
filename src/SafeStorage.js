import storageMock from './storageMock'
import InMemoryStorage from './inMemoryStorage'
import { getErrorMessage, storageTypes, validateStorageType } from './util'

export default class SafeStorage {
  static isNativeStorageSupported(storageType) {
    validateStorageType(storageType)

    const TEST_VALUE = 'test_value'

    try {
      const key = 'safe-web-storage.support'

      window[storageType].setItem(key, TEST_VALUE)
      const value = window[storageType].getItem(key)
      window[storageType].removeItem(key)

      return value === TEST_VALUE
    } catch (e) {
      return false
    }
  }

  constructor(storageType, options) {
    validateStorageType(storageType)

    this._options = {
      errorMessage: getErrorMessage(storageType),
      ...options,
    }

    const isSessionStorage = storageType === storageTypes.sessionStorage

    if (!SafeStorage.isNativeStorageSupported(storageType)) {
      this._storage = isSessionStorage
        ? InMemoryStorage
        : storageMock(this._options)
      this._isInMemoryStorageUsed = isSessionStorage
      this._isNativeStorageUsed = false

      console.warn(this._options.errorMessage)

      return
    }

    this._storage = window[storageType]
    this._storageType = storageType
    this._isNativeStorageUsed = true
    this._isInMemoryStorageUsed = false
  }

  get isNativeStorageUsed() {
    return this._isNativeStorageUsed
  }

  get isInMemoryStorageUsed() {
    return Boolean(this._isInMemoryStorageUsed)
  }
}
