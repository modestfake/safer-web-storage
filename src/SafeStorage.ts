import storageMock from './storageMock'
import InMemoryStorage from './inMemoryStorage'
import { getErrorMessage, StorageType, validateStorageType } from './util'

export interface Options {
  errorMessage?: string
}

export default class SafeStorage implements Storage {
  private _storageType: StorageType
  private _isNativeStorageUsed: boolean
  private _isInMemoryStorageUsed: boolean
  private _options: {
    errorMessage: string
  }
  private _storage: Storage

  static isNativeStorageSupported(storageType: StorageType) {
    validateStorageType(storageType)

    const TEST_VALUE = 'test_value'

    try {
      const key = 'safer-web-storage.support'

      window[storageType].setItem(key, TEST_VALUE)
      const value = window[storageType].getItem(key)
      window[storageType].removeItem(key)

      return value === TEST_VALUE
    } catch (e) {
      return false
    }
  }

  constructor(storageType: StorageType, options: Options = {}) {
    validateStorageType(storageType)

    this._storageType = storageType

    this._options = {
      errorMessage: getErrorMessage(storageType),
      ...options,
    }

    const isSessionStorage = storageType === 'sessionStorage'

    if (!SafeStorage.isNativeStorageSupported(storageType)) {
      this._storage = isSessionStorage
        ? InMemoryStorage
        : storageMock(this._options)
      this._isInMemoryStorageUsed = isSessionStorage
      this._isNativeStorageUsed = false

      console.warn(this._options.errorMessage)

      return
    }

    this._storage = window[storageType]
    this._isNativeStorageUsed = true
    this._isInMemoryStorageUsed = false
  }

  get isNativeStorageUsed() {
    return this._isNativeStorageUsed
  }

  get isInMemoryStorageUsed() {
    return Boolean(this._isInMemoryStorageUsed)
  }

  getItem(key: string): string | null {
    return this._storage.getItem(key)
  }

  setItem(key: string, value: string): void {
    return this._storage.setItem(key, value)
  }

  removeItem(key: string): void {
    return this._storage.removeItem(key)
  }

  key(index: number): string | null {
    return this._storage.key(index)
  }

  clear(): void {
    return this._storage.clear()
  }

  get length(): number {
    return this._storage.length
  }
}
