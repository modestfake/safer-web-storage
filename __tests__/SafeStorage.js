import storageMock from './helpers/__mocks__/storage'
import SafeStorage from '../src/SafeStorage'

const previousWindow = global.window

beforeEach(() => {
  window.localStorage = undefined
  window.sessionStorage = undefined
  console.warn = jest.fn()
})

afterEach(() => {
  jest.clearAllMocks()
  jest.resetModules()
  global.window = previousWindow
  jest.restoreAllMocks()
})

describe('SafeStorage', () => {
  test('Passed wrong storage type', () => {
    expect(() => {
      new SafeStorage()
    }).toThrow('Please provide one of storages: sessionStorage, localStorage')

    expect(() => {
      new SafeStorage('wrongStorage')
    }).toThrow(
      '"wrongStorage" is not supported. use one of these storages: sessionStorage, localStorage'
    )
  })

  test('Use native localStorage', () => {
    window.localStorage = storageMock
    const safeLocalStorage = new SafeStorage('localStorage')

    expect(SafeStorage.isNativeStorageSupported('localStorage')).toBe(true)
    expect(safeLocalStorage.isNativeStorageUsed).toBe(true)
    expect(safeLocalStorage.isInMemoryStorageUsed).toBe(false)
    expect(safeLocalStorage._options).toMatchSnapshot()
  })

  test('Use native sessionStorage', () => {
    window.sessionStorage = storageMock
    const safeSessionStorage = new SafeStorage('sessionStorage')

    expect(SafeStorage.isNativeStorageSupported('sessionStorage')).toBe(true)
    expect(safeSessionStorage.isNativeStorageUsed).toBe(true)
    expect(safeSessionStorage.isInMemoryStorageUsed).toBe(false)
  })

  test('Pass custom error message', () => {
    const errorMessage = 'Custom error message'
    const safeLocalStorage = new SafeStorage('localStorage', { errorMessage })

    expect(safeLocalStorage._options.errorMessage).toBe(errorMessage)
  })

  test('Handle AccessDenied error', () => {
    const errorMessage =
      'Uncaught DOMException: Failed to read the "localStorage" property from "Window": Access is denied for this document.'

    delete global.window

    global.window = new Proxy(previousWindow, {
      get(obj, prop) {
        if (['localStorage', 'sessionStorage'].includes(prop)) {
          throw new Error(errorMessage)
        }

        if (prop in obj) {
          return obj[prop]
        }

        return undefined
      },
    })

    const safeSessionStorage = new SafeStorage('sessionStorage')

    expect(safeSessionStorage.isInMemoryStorageUsed).toBe(true)
    expect(safeSessionStorage.isNativeStorageUsed).toBe(false)
    expect(SafeStorage.isNativeStorageSupported('sessionStorage')).toBe(false)

    const safeLocalStorage = new SafeStorage('localStorage')

    expect(safeLocalStorage.isInMemoryStorageUsed).toBe(false)
    expect(safeLocalStorage.isNativeStorageUsed).toBe(false)
    expect(SafeStorage.isNativeStorageSupported('localStorage')).toBe(false)
  })
})
