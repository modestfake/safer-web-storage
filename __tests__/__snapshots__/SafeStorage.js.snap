// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeStorage Use native localStorage 1`] = `
Object {
  "errorMessage": "Looks like you've disabled \\"localStorage\\". Enable it to avoid this warning.",
}
`;
